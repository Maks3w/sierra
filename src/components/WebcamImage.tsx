"use client"

import Image from 'next/image';
import { useEffect, useState } from 'react';
import {Webcam} from "@/config/places";

interface WebcamImageProps {
    webcam: Webcam,
    className?: string,
}

const WebcamImage = ({ webcam, className }: WebcamImageProps) => {
    const [loading, setLoading] = useState(true);
    const [error, setError] = useState<string | null>(null);
    const [imageUrl, setImageUrl] = useState(`${webcam.url}?${new Date().getTime()}`);

    const handleImageLoad = () => {
        setLoading(false);
        setError(null);
    };

    const handleImageError = () => {
        setLoading(false);
        setError('Failed to load image');
    };

    useEffect(() => {
        const interval = setInterval(refreshHandler, webcam.refreshInterval * 60 * 1000);
        return () => clearInterval(interval);
    }, [webcam]);

    const refreshHandler = () => {
        setLoading(true);
        const newUrl = `${webcam.url}?${new Date().getTime()}`;
        setImageUrl(newUrl);
    };

    return (
        <div data-testid="webcam-image">
            {loading && (
                <div data-testid="webcam_image-loader" className="absolute inset-0 flex items-center justify-center bg-gray-800 bg-opacity-75">
                    <div className="loader"></div>
                </div>
            )}
            <>
                <Image
                    src={imageUrl}
                    alt={webcam.name}
                    width={800}
                    height={600}
                    className={className}
                    loading="lazy"
                    onLoad={handleImageLoad}
                    onError={handleImageError}
                    unoptimized
                />
                {error && <p className="text-center text-red-500">{error}</p>}
            </>
        </div>
    );
};

export default WebcamImage;