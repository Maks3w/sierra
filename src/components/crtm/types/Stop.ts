export interface GetStopsInformationResponse {
  stops: { StopInformation: Stop };
}

export interface Stop {
  codStop: string;
  name: string;
  lines: Line[];
  location: {
    lat: number;
    lng: number;
  };
  services: string[];
}

interface Line {
  lineId: string;
  destination: string;
  time: string;
}

export interface StopTimes {
  actualDate: string;
  stop: {
    codStop: string;
    shortCodStop: string;
    name: string;
    park: number;
    nightLinesService: number;
  };
  times: {
    Time?: Time[];
  };
  linesStatus: {
    LineStatus: LineStatus[];
  };
}

interface Time {
  line: {
    codLine: string;
    shortDescription: string;
    description: string;
    codMode: string;
    updateDate: string;
    updateKmlDate: string;
    nightService: number;
    active: boolean;
    shortItinerary: Record<string, unknown>;
    companyCode: string;
  };
  direction: number;
  destination: string;
  destinationStop: {
    codStop: string;
    shortCodStop: string;
    name: string;
    park: number;
    nightLinesService: number;
  };
  time: string;
  codVehicle: string;
  codIssue: string;
}

interface LineStatus {
  line: {
    codLine: string;
    shortDescription: string;
  };
  SAEStatus: boolean;
}

export interface GetStopsTimesResponse {
  stopTimes: StopTimes;
}