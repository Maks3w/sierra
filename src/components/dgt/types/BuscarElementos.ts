export interface BuscarElementosParams {
  latNS: number;
  longNS: number;
  latSW: number;
  longSW: number;
  zoom: number;
  accion?: string;
  Camaras?: boolean;
  SensoresTrafico?: boolean;
  SensoresMeteorologico?: boolean;
  Paneles?: boolean;
  Radares?: boolean;
  IncidenciasRETENCION?: boolean;
  IncidenciasOBRAS?: boolean;
  IncidenciasMETEOROLOGICA?: boolean;
  IncidenciasPUERTOS?: boolean;
  IncidenciasOTROS?: boolean;
  IncidenciasEVENTOS?: boolean;
  IncidenciasRESTRICCIONES?: boolean;
  niveles?: boolean;
  caracter?: string;
}

export interface Elemento {
  tipo: ElementoTipo;
  estado: number;
  lng: number;
  sentido: string;
  alias: string;
  carretera: string;
  PK: number;
  codEle: string;
  lat: number;
  descripcion?: string;
  icono?: string;
  autonomia?: string;
  causa?: string;
  hora?: string;
  precision?: string;
  pkFinal?: number;
  provincia?: string;
  fechaFin?: string;
  horaFin?: string;
  fecha?: string;
  pkIni?: number;
  poblacion?: string;
  suceso?: string;
  nivel?: string;
}

export enum ElementoTipo {
  SensorMeteorologico = 'SensorMeteorologico',
  Panel_CMS = 'Panel_CMS',
  Camara = 'Camara',
  Incidencia = 'Incidencia',
}

export const paramsDefaults = {
  accion: 'getElementos',
  Camaras: true,
  SensoresTrafico: true,
  SensoresMeteorologico: true,
  Paneles: true,
  Radares: true,
  IncidenciasRETENCION: true,
  IncidenciasOBRAS: true,
  IncidenciasMETEOROLOGICA: true,
  IncidenciasPUERTOS: true,
  IncidenciasOTROS: true,
  IncidenciasEVENTOS: true,
  IncidenciasRESTRICCIONES: true,
  niveles: true,
  caracter: 'acontecimiento'
}
