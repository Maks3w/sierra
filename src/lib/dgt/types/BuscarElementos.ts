export interface BuscarElementosParams {
  latNS: number;
  longNS: number;
  latSW: number;
  longSW: number;
  zoom?: number;
  accion?: string;
  Camaras?: boolean;
  SensoresTrafico?: boolean;
  SensoresMeteorologico?: boolean;
  Paneles?: boolean;
  Radares?: boolean;
  IncidenciasRETENCION?: boolean;
  IncidenciasOBRAS?: boolean;
  IncidenciasMETEOROLOGICA?: boolean;
  IncidenciasPUERTOS?: boolean;
  IncidenciasOTROS?: boolean;
  IncidenciasEVENTOS?: boolean;
  IncidenciasRESTRICCIONES?: boolean;
  niveles?: boolean;
  caracter?: string;
}

export interface Elemento {
  tipo: ElementoTipo;
  estado: number;
  lng: number;
  sentido: string;
  alias: string;
  carretera: string;
  PK: number;
  codEle: string;
  lat: number;
  descripcion?: string;
  icono?: string;
  autonomia?: string;
  causa?: string;
  hora?: string;
  precision?: string;
  pkFinal?: number;
  provincia?: string;
  fechaFin?: string;
  horaFin?: string;
  fecha?: string;
  pkIni?: number;
  poblacion?: string;
  suceso?: string;
  nivel?: string;
  encodedPoints?: string;
  encodedLevels?: string;
  textoAdvertenciaPrecision?: string;
  indiceMapa?: string;
}

export enum ElementoTipo {
  SensorMeteorologico = 'SensorMeteorologico',
  Panel_CMS = 'Panel_CMS',
  Camara = 'Camara',
  Incidencia = 'Incidencia',
  NivelServicio = 'nivelServicio',
  Radar = 'Radar',
  SensorTrafico = 'SensorTrafico'
}

export const paramsDefaults = {
  zoom: 18,
  accion: 'getElementos',
  Camaras: true,
  SensoresTrafico: true,
  SensoresMeteorologico: true,
  Paneles: true,
  Radares: true,
  IncidenciasRETENCION: true,
  IncidenciasOBRAS: true,
  IncidenciasMETEOROLOGICA: true,
  IncidenciasPUERTOS: true,
  IncidenciasOTROS: true,
  IncidenciasEVENTOS: true,
  IncidenciasRESTRICCIONES: true,
  niveles: true,
  caracter: 'acontecimiento'
}

export interface Panel_CMS {
  mensaje2: string;
  mensaje1: string;
  imgTxtIzq1: string;
  imgTxtIzq2: string;
  textoAdvertenciaPrecision: string;
  indiceMapa: string;
  drawDer2: string;
  drawIz2: string;
  tipo: string;
  drawDer1: string;
  drawIz1: string;
  imgTxtDer2: string;
  imgTxtDer1: string;
  NumAlternancias: string;
}

export interface SensorMeteorologico {
  t_congel: string;
  tipo: string;
  vel_viento: string;
  dir_viento: string;
  est_super: string;
  t_subsuelo: string;
  textoAdvertenciaPrecision: string;
  tip_viento: string;
  alt_agua: string;
  t_super: string;
  tiempo_Presente: string;
  visibilidad: string;
  radiacion_global: string;
  fecha: string;
  tipo_viento: string;
  temp_rocio: string;
  c_Precipitaciones: string;
  i_Precipitaciones: string;
  temperatura: string;
  humedad: string;
  n_Precipitacion: string;
  salinidad: string;
  presion_A: string;
  indiceMapa: string;
  noDatos?: string;
}

export const sensorMeteorologicoLabels: { [key: string]: string } = {
  fecha: "Fecha",
  t_congel: "Temperatura de Congelación",
  vel_viento: "Velocidad del Viento",
  dir_viento: "Dirección del Viento",
  est_super: "Estado de la Superficie",
  t_subsuelo: "Temperatura del Subsuelo",
  tip_viento: "Tipo de Viento",
  alt_agua: "Altura del Agua",
  t_super: "Temperatura de la Superficie",
  tiempo_Presente: "Tiempo Presente",
  visibilidad: "Visibilidad",
  radiacion_global: "Radiación Global",
  tipo_viento: "Tipo de Viento",
  temp_rocio: "Temperatura del Rocío",
  c_Precipitaciones: "Precipitaciones",
  i_Precipitaciones: "Intensidad de Precipitaciones",
  temperatura: "Temperatura",
  humedad: "Humedad",
  n_Precipitacion: "Nivel de Precipitación",
  salinidad: "Salinidad",
  presion_A: "Presión Atmosférica",
};

export interface SensorTrafico {
  fecha: string;
  tipo: string;
  textoAdvertenciaPrecision: string;
  intensidad: string;
  ocupacion: string;
  composicion: string;
  velocidad: string;
  indiceMapa: string;
  noDatos: string;
}

export interface Incidencia {
  icono: string;
  descripcion: string;
  tipo: string;
  estado: number;
  autonomia: string;
  lng: number;
  causa: string;
  hora: string;
  precision: string;
  pkFinal: number;
  provincia: string;
  fechaFin: string;
  horaFin: string;
  fecha: string;
  pkIni: number;
  sentido: string;
  alias: string;
  carretera: string;
  poblacion: string;
  suceso: string;
  codEle: string;
  nivel: string;
  lat: number;
}
