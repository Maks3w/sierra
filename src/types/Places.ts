export interface Webcam {
    name: string;
    providerName: string;
    providerImage: string;
    refreshInterval: number;
    delayTime: number;
    url: string;
}

export interface SocialNetwork {
    twitter?: string;
}

export enum PublicTransportType {
    BUS = "bus",
    TRAIN = "train",
}

export enum PublicTransportDirection {
    BOTH = "both",
    UP = "up",
    DOWN = "down",
}

export interface PublicTransport {
    type: PublicTransportType;
    id: string;
    direction: string;
}

export interface Weather {
  type: string;
  value: string;
}

export interface Place {
    name: string;
    webcams: Webcam[];
    infoNieve?: number;
    socialNetwork: SocialNetwork;
    publicTransport: PublicTransport[];
    weather: Weather[];
}
