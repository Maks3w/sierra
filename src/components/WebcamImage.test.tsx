import {fireEvent, render, screen} from '@testing-library/react';
import WebcamImage from './WebcamImage';
import { Webcam } from '@/types/Places';

const mockWebcam: Webcam = {
  name: 'Test Webcam',
  url: 'https://example.com/webcam.jpg',
  refreshInterval: 1,
  providerName: 'Test Provider',
  providerImage: 'My image',
  delayTime: 0,
};

beforeEach(() => {
  vi.useFakeTimers();
  vi.setSystemTime(new Date('2023-01-01T00:00:00Z'));
});

afterEach(() => {
  vi.useRealTimers();
});

describe('WebcamImage', () => {
  it('shows loading state initially', () => {
    render(<WebcamImage webcam={mockWebcam} />);
    const component = screen.getByTestId("webcam-image");
    expect(component).toMatchSnapshot();
  });

  it('handles image error', async () => {
    render(<WebcamImage webcam={mockWebcam} />);
    const image = screen.getByAltText(mockWebcam.name);

    fireEvent.error(image)
    const component = screen.getByTestId("webcam-image");
    expect(component).toMatchSnapshot();
  });
})
