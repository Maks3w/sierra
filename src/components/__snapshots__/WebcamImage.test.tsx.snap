// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`WebcamImage > handles image error 1`] = `
<div
  data-testid="webcam-image"
>
  <img
    alt="Test Webcam"
    data-nimg="1"
    decoding="async"
    height="600"
    loading="lazy"
    src="https://example.com/webcam.jpg?1672531200000"
    style=""
    width="800"
  />
  <p
    class="text-center text-red-500"
  >
    Failed to load image
  </p>
</div>
`;

exports[`WebcamImage > shows loading state initially 1`] = `
<div
  data-testid="webcam-image"
>
  <div
    class="absolute inset-0 flex items-center justify-center bg-gray-800 bg-opacity-75"
    data-testid="webcam_image-loader"
  >
    <div
      class="loader"
    />
  </div>
  <img
    alt="Test Webcam"
    data-nimg="1"
    decoding="async"
    height="600"
    loading="lazy"
    src="https://example.com/webcam.jpg?1672531200000"
    style="color: transparent;"
    width="800"
  />
</div>
`;
